core/src/main/java/com/yahoo/imapnio/async/response/ImapResponseMapper.java

```diff
@@ -442,18 +477,116 @@ private SearchResult parseToSearchResult(@Nonnull final IMAPResponse[] ir) throw
             }
             final List<Long> v = new ArrayList<Long>(); // will always return a non-null array
 
+            Long modSeq = null;
+
             // Grab all SEARCH responses
             long num;
             for (final IMAPResponse sr : ir) {
                 // There *will* be one SEARCH response.
                 if (sr.keyEquals("SEARCH")) {
                     while ((num = sr.readLong()) != -1) {
-                        v.add(Long.valueOf(num));
+                        v.add(num);
+                    }
+                    if (sr.readByte() == '(') {
+                        final String s = sr.readAtom();
+                        if (s.equalsIgnoreCase("MODSEQ")) {
+                            modSeq = sr.readLong();
+                        }
                     }
                 }
             }
 
-            return new SearchResult(v);
+            return new SearchResult(v, modSeq);
+        }
+
+        /**
+         * Parses the responses from Store command and UID Store command to a {@link StoreResult} object.
+         *
+         * @param ir the list of responses from Store or UID Store command, the input responses array should contain the tagged/final one
+         * @param extensionItems the array of extension FetchItem
+         * @return StoreResult object constructed based on the given IMAPResponse array,
+         * @throws ImapAsyncClientException when tagged response is bad or given response length is 0
+         * @throws ProtocolException when fetch response cannot be parsed
+         */
+        @Nonnull
+        private StoreResult parseToStoreResult(@Nonnull final IMAPResponse[] ir, @Nonnull final FetchItem[] extensionItems)
+                throws ImapAsyncClientException, ProtocolException {
+            if (ir.length < 1) {
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final Response taggedResponse = ir[ir.length - 1];
+            if (taggedResponse.isBAD()) {
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final List<FetchResponse> fetchResponses = new ArrayList<>(); // will always return a non-null array
+
+            MessageNumberSet[] modifiedMsgsets = null; // only one response will contain modified numbers
+            Long highestModSeq = null;
+
+            for (final IMAPResponse sr: ir) {
+                sr.skipSpaces();
+                if (sr.keyEquals("FETCH")) {
+                    fetchResponses.add(parseToFetchResponse(sr, extensionItems));
+                } else if (sr.readByte() == (byte) L_BRACKET) { // HIGHESTMODSEQ or MODIFIED
+                    final String responseCode = sr.readAtom();
+                    if (sr.isOK() && responseCode.equalsIgnoreCase("HIGHESTMODSEQ")) {
+                        highestModSeq = sr.readLong();
+                    } else if (responseCode.equalsIgnoreCase("MODIFIED")) {
```

Comment:
shouldnt this be in tagged response?  can u put the abnf here?

Reply:
resp-text-code      =/ "HIGHESTMODSEQ" SP mod-sequence-value /
                          "NOMODSEQ" /
                          "MODIFIED" SP sequence-set

Reply:
As discussed previously, this allows No response, so we check whether it is bad in the beginning.